crypto/fips/self_test.c

```diff
@@ -0,0 +1,349 @@
+            || memcmp(decoded, plaintxt,  decoded_len) != 0) {
+        goto err;
+    }
+    ret = 0;
+err:
+    SELF_TEST_EVENT_onend(&ev, ret);
+    OPENSSL_free(ciphertxt);
+    return ret;
+}
+
+int SELF_TEST_keygen_pairwise_test_ecdsa(EC_KEY *eckey)
+{
+    int ret = 1;
+    unsigned char dgst[16] = {0};
+    unsigned int dgst_len = (unsigned int)sizeof(dgst);
+    ECDSA_SIG *sig = NULL;
+    ST_EVENT ev;
+
+    SELF_TEST_EVENT_init(&ev);
+    SELF_TEST_EVENT_onbegin(&ev, SELF_TEST_TYPE_PCT, SELF_TEST_DESC_PCT_ECDSA);
+
+    sig = ECDSA_do_sign(dgst, dgst_len, eckey);
+    if (sig == NULL)
+        goto err;
+
+    SELF_TEST_EVENT_oncorrupt_byte(&ev, dgst);
+
+    if (ECDSA_do_verify(dgst, dgst_len, sig, eckey) != 1)
+        goto err;
+
+    ret = 0;
+err:
+    SELF_TEST_EVENT_onend(&ev, ret);
+    ECDSA_SIG_free(sig);
+    return ret;
+}
+
+int SELF_TEST_keygen_pairwise_test_dsa(DSA *dsa)
+{
+    int ret = 1;
+    unsigned char dgst[16] = {0};
+    unsigned int dgst_len = (unsigned int)sizeof(dgst);
+    DSA_SIG *sig = NULL;
+    ST_EVENT ev;
+
+    SELF_TEST_EVENT_init(&ev);
+    SELF_TEST_EVENT_onbegin(&ev, SELF_TEST_TYPE_PCT, SELF_TEST_DESC_PCT_DSA);
+
+    sig = DSA_do_sign(dgst, (int)dgst_len, dsa);
+    if (sig == NULL)
+        goto err;
+
+    SELF_TEST_EVENT_oncorrupt_byte(&ev, dgst);
+
+    if (DSA_do_verify(dgst, dgst_len, sig, dsa) != 1)
+        goto err;
+
+    ret = 0;
+err:
+    SELF_TEST_EVENT_onend(&ev, ret);
+    DSA_SIG_free(sig);
+    return ret;
+}
+
+#define INTEGRITY_BUF_SIZE (4 * 1024)
+
+/*
+ * Calculate the HMAC SHA256 of data read using a BIO and read_cb, and verify
+ * the result matches the expected value.
+ * Return 1 if verified, or 0 if it fails.
+ */
+static int verify_integrity(BIO *bio, BIO_READ_CB read_cb,
+                            unsigned char *expected, size_t expected_len,
+                            ST_EVENT *ev, int type)
+{
+    int ret = 0;
+    HMAC_CTX *ctx = NULL;
+    unsigned char out[EVP_MAX_MD_SIZE];
+    unsigned int out_len = 0;
+    size_t bytes_read = 0;
+    unsigned char buf[INTEGRITY_BUF_SIZE];
+    const EVP_MD *md = EVP_sha256();
+
+    SELF_TEST_EVENT_onbegin(ev, type, SELF_TEST_DESC_INTEGRITY_HMAC);
+
+    /* TODO - Is HMAC_ the correct interface to call ? */
+    ctx = HMAC_CTX_new();
+    if (ctx == NULL)
+        goto err;
+
+    if (!HMAC_Init_ex(ctx, fixed_key, (int)sizeof(fixed_key), md, NULL))
+        goto err;
+    while (1) {
+        ret = read_cb(bio, buf, sizeof(buf), &bytes_read);
+        if (ret != 1)
+            break;
+        if (!HMAC_Update(ctx, buf, bytes_read))
+            goto err;
+    }
+    if (!HMAC_Final(ctx, out, &out_len))
+        goto err;
+
+    SELF_TEST_EVENT_oncorrupt_byte(ev, out);
+
+    if (expected_len != (size_t)out_len
+            || memcmp(expected, out, (size_t)out_len) != 0)
+        goto err;
+    ret = 0;
+err:
+    SELF_TEST_EVENT_onend(ev, ret);
+    HMAC_CTX_free(ctx);
+    return ret;
+}
+
+#if 0
```

Comment:
?

Reply:
There will be lots of #if 0 in here soon..
Most of the code will rely on there being EVP_ methods available in the provider..
For now it will just be an empty shell.